src/object/srv_obj.c

```diff
@@ -41,6 +41,299 @@
+
+		dta = ds_tx_prepare_args(cont_hdl, cont, false);
+		if (dta == NULL)
+			D_WARN("Not enough RAM for DTX aggregation on "
+			       DF_UUID"\n", DP_UUID(cont->sc_uuid));
+		else
+			dss_ult_create(ds_tx_aggregate, dta, DSS_ULT_SELF,
+				dss_get_module_info()->dmi_tgt_id, 0, NULL);
+		break;
+	}
+	case DTX_ACT_COMMIT_ASYNC:
+		if (!cont->sc_tx_committing) {
+			struct ds_tx_args *dta;
+
+			dta = ds_tx_prepare_args(cont_hdl, cont, false);
+			if (dta == NULL) {
+				D_WARN("Not enough RAM for DTX commit on "
+				       DF_UUID"\n", DP_UUID(cont->sc_uuid));
+			} else {
+				cont->sc_tx_committing = 1;
+				rc = dss_ult_create(ds_tx_commit, dta,
+					DSS_ULT_SELF,
+					dss_get_module_info()->dmi_tgt_id,
+					0, NULL);
+				if (rc != 0) {
+					D_WARN("Failed to trigger async DTX "
+					       "commit on "DF_UUID": rc = %d\n",
+					       DP_UUID(cont->sc_uuid), rc);
+					cont->sc_tx_committing = 0;
+				}
+			}
+		}
+		break;
+	case DTX_ACT_COMMIT_SYNC: {
+		struct daos_tx_entry	 dte;
+
+		dte.dte_xid = *dti;
+		dte.dte_oid = *oid;
+		/* Commit current DTX synchronously. */
+		rc = dtx_commit(cont_hdl->sch_pool->spc_uuid, cont->sc_uuid,
+				&dte, 1, cont_hdl->sch_pool->spc_map_version);
+		if (rc != 0)
+			err = -DER_INPROGRESS;
+		break;
+	}
+	default:
+		if (err < 0)
+			ds_tx_abort(cont_hdl->sch_pool->spc_uuid,
+				    cont->sc_uuid, dti, oid,
+				    cont_hdl->sch_pool->spc_map_version);
+		break;
+	}
+
+	return err < 0 ? err : 0;
+}
+
+static int
+obj_prepare_cos(struct ds_cont *cont, struct ds_cont_hdl *cont_hdl,
+		daos_unit_oid_t *oid, daos_key_t *dkey,
+		uint32_t types, struct daos_tx_id **dtis)
+{
+	int	rc = 0;
+
+	if (!srv_enable_dtx)
+		return 0;
+
+	do {
+		rc = vos_dtx_list_cos(cont->sc_hdl, oid, dkey, types, dtis);
+		if (rc < 0)
+			return rc;
+
+		/* There are too many shared DTXs to be committed,
+		 * as to cannot be taken via the normal UPDATE RPC.
+		 * Then needs to commit them firstly.
+		 */
+		if (rc > 0 && *dtis == NULL) {
+			struct ds_tx_args *dta;
+
+			dta = ds_tx_prepare_args(cont_hdl, cont, true);
+			if (dta == NULL) {
+				D_ERROR("Not enough RAM for DTX commit.\n");
+				return -DER_NOMEM;
+			}
+
+			if (cont->sc_tx_committing)
+				/* Someone is aggregated committing DTXs,
+				 * then let's ask the client to retry some
+				 * time later instead of making current ULT
+				 * to be blocked.
+				 */
+				return -DER_INPROGRESS;
+
+			cont->sc_tx_committing = 1;
+			ds_tx_commit(dta);
+		}
+	} while (rc > 0 && *dtis == NULL);
+
+	return rc;
+}
+
+static int
+obj_handle_resend(daos_handle_t coh, daos_unit_oid_t *oid,
+		  struct daos_tx_id *dti, daos_key_t *dkey, bool punch)
+{
+	uint64_t	hash = 0;
+	int		rc;
+
+	if (dkey != NULL)
+		hash = d_hash_murmur64(dkey->iov_buf, dkey->iov_len,
+				       VOS_BTR_MUR_SEED);
+	rc = vos_dtx_check_committable(coh, oid, dti, hash, punch);
+	switch (rc) {
+	case DTX_ST_PREPARED:
+		return 0;
+	case DTX_ST_INIT:
+		if (dti->dti_sec >= srv_start_time)
```

Comment:
this time compare seems un-trustable, even use HLC later the server local start time HLC also cannot be used to  compared with client-side RPC's HLC timestamp.

resend when DTX_ST_INIT seems very rarely, maybe can just abort as below is enough.

Reply:
Seems fine to compare the RPC receive HLC with server start HLC, to check if the RPC was send after srv restart. dti_sec seems not quite useful for this.

Reply:
Current time(NULL) based resend handling is not accurate, that is the temporary solution before HLC introduced. The DTX ID may be totally replaced by the HLC in the future.

Reply:
Before the HLC introduce, the most potential trouble the clock drift between client and server when the server restart. But since we do not support server re-integration yet, it is not a big issue now.